### ISSUE TITLE
Weight decay modifies grad for SGD but not for Adam

### ISSUE THREAD

**Issue by bartwojcik** [pytorch/pytorch#12218]
Seems that in step() of SGD optimizer:
```
                d_p = p.grad.data
                if weight_decay != 0:
                    d_p.add_(weight_decay, p.data)
```
gradient is being modified. Yet, Adam does a copy:
```
                if group['weight_decay'] != 0:
                    grad = grad.add(group['weight_decay'], p.data)
```
This behavior should be documented for both optimizers.

I find it very unintuitive and surprising that gradients can be modified in simple, single step (the one without reevaluation closure), but I guess there could be some reason and that is intended behavior so I am filing this under documentation.

**Comment by kashif** [comment]
no adam modifies it in place see:

https://github.com/pytorch/pytorch/blob/master/torch/optim/adam.py#L90

**Comment by ngimel** [comment]
Closing due to age
